### PR TITLE
APEXMALHAR-1953 Updated Jdbc Output Operator

### PR DESCRIPTION
**Pull request by bhupeshchawda**
1. Renaming JdbcPOJOOutputOperator to AbstractJdbcPOJOOutputOperator.
2. Added JdbcPOJOInsertOutputOperator for insert queries. Added support for automatic mapping of fields from POJO to DB columns. 
3. Added JdbcPOJONonInsertOutputOperator for update / delete / merge queries. Refactored accordingly. 
   Added unit tests.

### CODE REVIEW

File: library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcPOJOInsertOutputOperator.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+  @Override
+  public void setup(OperatorContext context)
+  {
+    super.setup(context);
+
+    // Populate columnNames and columnDataTypes
+    try {
+      if (getFieldInfos() == null) { // then assume direct mapping
+        LOG.info("Assuming direct mapping between POJO fields and DB columns");
```
Comment (sandeepdeshmukh) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcPOJOInsertOutputOperator.java]: Improve the log message to include explicit message about missing field info.

File: library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcPOJOInsertOutputOperator.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+      Field[] fields = pojoClass.getDeclaredFields();
+      // Create fieldInfos in case of direct mapping
+      List<JdbcFieldInfo> fieldInfos = Lists.newArrayList();
+      for (int i = 0; i < columnNames.size(); i++) {
+        String columnName = columnNames.get(i);
+        String pojoField = getMatchingField(fields, columnName);
+
+        if (columnNullabilities.get(i) == ResultSetMetaData.columnNoNulls &&
+            (pojoField == null || pojoField.length() == 0)) {
+          throw new RuntimeException("Data for a non-nullable field not found in POJO");
```
Comment (sandeepdeshmukh) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcPOJOInsertOutputOperator.java]: Add field name in the message.

File: library/src/test/java/com/datatorrent/lib/db/jdbc/JdbcOperatorTest.java [hunk 3]
```diff
@@ -332,7 +400,154 @@ public void testJdbcPojoOutputOperator()
...
+    for (TestPOJOEvent event : events) {
+      outputOperator.input.process(event);
+    }
+    outputOperator.endWindow();
+
+    Assert.assertEquals("rows in db", 10, outputOperator.getNumOfEventsInStore(TABLE_POJO_NAME_NAME_DIFF));
+    Assert.assertEquals("null name rows in db", 10,
+        outputOperator.getNumOfNullEventsInStore(TABLE_POJO_NAME_NAME_DIFF));
+  }
+
```
Comment (sandeepdeshmukh) [inline library/src/test/java/com/datatorrent/lib/db/jdbc/JdbcOperatorTest.java:512]: Add negative test: non-nullable filed in DB is null in POJO.

File: library/src/test/java/com/datatorrent/lib/db/jdbc/JdbcOperatorTest.java [hunk 4]
```diff
@@ -332,7 +400,154 @@ public void testJdbcPojoOutputOperator()
     }
     outputOperator.endWindow();
 
-    Assert.assertEquals("rows in db", 10, outputOperator.getNumOfEventsInStore());
+    Assert.assertEquals("rows in db", 10, outputOperator.getNumOfEventsInStore(TABLE_POJO_NAME));
+  }
+
+  /**
+   * This test will assume direct mapping for POJO fields to DB columns
+   */
```
Comment (sandeepdeshmukh) [inline library/src/test/java/com/datatorrent/lib/db/jdbc/JdbcOperatorTest.java:410]: Add list of scenarios you would like to cover.

File: library/src/test/java/com/datatorrent/lib/db/jdbc/JdbcOperatorTest.java [hunk 5]
```diff
@@ -332,7 +400,154 @@ public void testJdbcPojoOutputOperator()
...
+    }
+
+    updateOperator.getDistinctNonUnique();
+    updateOperator.beginWindow(0);
+    for (TestPOJOEvent event : events) {
+      updateOperator.input.process(event);
+    }
+    updateOperator.endWindow();
+
+    Assert.assertEquals("rows in db", 10, updateOperator.getNumOfEventsInStore());
```
Comment (sandeepdeshmukh) [inline library/src/test/java/com/datatorrent/lib/db/jdbc/JdbcOperatorTest.java:608]: Add comment on how this is verified.

**Comment by sandeepdeshmukh** [comment]
Would be nice to have an error port where you can redirect all erroneous tuples.

File: library/src/test/java/com/datatorrent/lib/db/jdbc/JdbcOperatorTest.java (hunk 3 above)
Comment (bhupeshchawda) [inline library/src/test/java/com/datatorrent/lib/db/jdbc/JdbcOperatorTest.java:512]: Added

**Comment by bhupeshchawda** [comment]
Addressed review comments.

**Comment by DT-Priyanka** [comment]
Should we add any metrics to tuples processed or error tuples per window?

**Comment by bhupeshchawda** [comment]
Sure, will add metrics

**Comment by sandeepdeshmukh** [comment]
Looks good to me. Will merge in a day if there are no further comments.

**Comment by sandeepdeshmukh** [comment]
Bhupesh: Could you check following
2016-05-03 11:33:08,312 [main] INFO  fs.AbstractFileOutputOperator load - opened single.txt.2, active target/org.apache.apex.malhar.lib.fs.HDFSOutputOperatorTest/testSingleRollingFileCompletedWriteOverwriteInitial/single.txt.2
Tests run: 56, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 5.009 sec
Results :
Tests in error: 
  ManagedStateImplTest.testFreeWindowTransferRaceCondition:200 » NullPointer

Tests run: 902, Failures: 0, Errors: 1, Skipped: 0

The test failed on my laptop twice. Is there any race around condition? Is that related to your checkin?

File: library/src/main/java/com/datatorrent/lib/db/AbstractTransactionableStoreOutputOperator.java [hunk 6]
```diff
@@ -67,6 +68,8 @@ public void process(T t)
 
   };
 
+  public final transient DefaultOutputPort<T> error = new DefaultOutputPort<>();
```
Comment (otterc) [inline library/src/main/java/com/datatorrent/lib/db/AbstractTransactionableStoreOutputOperator.java:71]: Why is this port needed here?

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPOJOOutputOperator.java [hunk 7]
```diff
@@ -63,22 +58,18 @@
  * @since 2.1.0
  */
 @org.apache.hadoop.classification.InterfaceStability.Evolving
-public class JdbcPOJOOutputOperator extends AbstractJdbcTransactionableOutputOperator<Object>
-    implements Operator.ActivationListener<OperatorContext>
+public abstract class AbstractJdbcPOJOOutputOperator extends AbstractJdbcTransactionableOutputOperator<Object>
 {
-  @NotNull
-  private List<FieldInfo> fieldInfos;
+  private List<JdbcFieldInfo> fieldInfos;
```
Comment (otterc) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPOJOOutputOperator.java:63]: Can this field be null?

Comment (otterc) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPOJOOutputOperator.java:63]: Can this be called SqlFieldInfo

**Comment by otterc** [comment]
Please minimize the number of commits ensuring that rename doesn't lead to attribution loss.

File: library/src/main/java/com/datatorrent/lib/db/AbstractTransactionableStoreOutputOperator.java (hunk 6 above)
Comment (bhupeshchawda) [inline library/src/main/java/com/datatorrent/lib/db/AbstractTransactionableStoreOutputOperator.java:71]: This was added to track the error tuples which fail to be processed  / inserted into database. It made sense to add it at the same place where the input port was declared.

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPOJOOutputOperator.java (hunk 7 above)
Comment (bhupeshchawda) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPOJOOutputOperator.java:63]: Yes, fieldInfos can be null if the statement is an insert statement. In that case, it is generated by the operator code. I have inserted checks for this in the concrete POJO output operator.

Comment (bhupeshchawda) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPOJOOutputOperator.java:63]: We can call it SqlFieldInfo.
I think JdbcFieldInfo is also appropriate right?

File: library/src/main/java/com/datatorrent/lib/db/AbstractTransactionableStoreOutputOperator.java (hunk 6 above)
Comment (otterc) [inline library/src/main/java/com/datatorrent/lib/db/AbstractTransactionableStoreOutputOperator.java:71]: Is this port being used here? 
If not then why add it here

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPOJOOutputOperator.java (hunk 7 above)
Comment (otterc) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPOJOOutputOperator.java:63]: Isn't the list `fieldInfos` initialized to empty list in the constructor?

File: library/src/main/java/com/datatorrent/lib/db/AbstractTransactionableStoreOutputOperator.java (hunk 6 above)
Comment (bhupeshchawda) [inline library/src/main/java/com/datatorrent/lib/db/AbstractTransactionableStoreOutputOperator.java:71]: Moving the port to AbstractJdbcTransactionableOutputOperator

**Comment by bhupeshchawda** [comment]
Closing this PR due to improper branch name. Will open another on with the review comments addressed.
